### PR TITLE
executor: refine HashAgg.Close when unparallelExec

### LINKED ISSUE #8787: potential goroutine leak on parallel hash aggregate
```go
196 // Close implements the Executor Close interface.
197 func (e *HashAggExec) Close() error {
198     if e.isUnparallelExec {
199         e.childResult = nil
200         e.groupSet = nil
201         e.partialResultMap = nil
202         return nil
203     }
204     // `Close` may be called after `Open` without calling `Next` in test.
205     if !e.prepared {
206         close(e.inputCh)
207         for _, ch := range e.partialOutputChs {
208             close(ch)
209         }
210         close(e.finalOutputCh)
211     }
212     close(e.finishCh)
213     for _, ch := range e.partialOutputChs {
214         for range ch {
215         }
216     }
217     for range e.finalOutputCh {
218     }
219     return errors.Trace(e.baseExecutor.Close())
220 }
```

at line 202, we should call `child.Close()` before `return` to avoid goroutine leak.

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
fix #8787 

### What is changed and how it works?
Close the child executor when HashAgg executes unparalleled.

### Check List 

Tests 

exists leak tests

Code changes

 - Has exported function/method change

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

release-2.1

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8810)

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu Could you add a gofail test?

**Comment by shenli** [comment]
Please add a test case.

File: executor/builder.go [hunk 1]
```diff
@@ -1031,8 +1031,8 @@ func (b *executorBuilder) buildProjBelowAgg(aggFuncs []*aggregation.AggFuncDesc,
 	}
 
 	return &ProjectionExec{
-		baseExecutor: newBaseExecutor(b.ctx, expression.NewSchema(projSchemaCols...), projFromID, src),
-		//numWorkers:    b.ctx.GetSessionVars().ProjectionConcurrency,
```
Comment (winoros) [inline executor/builder.go]: Why change happened?

Comment (eurekaka) [reply executor/builder.go]: I commented out this line in https://github.com/pingcap/tidb/pull/8618 to pass leak test.

Comment (winoros) [reply executor/builder.go]: got

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason gofail test added

File: executor/seqtest/seq_executor_test.go [hunk 2]
```diff
@@ -642,6 +642,50 @@ func (s *seqTestSuite) TestIndexDoubleReadClose(c *C) {
 	atomic.StoreInt32(&executor.LookupTableTaskChannelSize, originSize)
 }
 
+func (s *seqTestSuite) TestParallelHashAggClose(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec(`use test;`)
+	tk.MustExec(`drop table if exists t;`)
+	tk.MustExec("create table t(a int, b int)")
+	tk.MustExec("insert into t values(1,1),(2,2)")
+	result := tk.MustQuery("desc select sum(a) from (select cast(t.a as signed) as a, b from t) t group by b;")
```
[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply executor/seqtest/seq_executor_test.go]: 1. The case can also check whether the child of HashAgg is closed correctly, so I make a Projection there.
2. Yep, the explain result is not really needed, I'll make it as a comment.

**Pull request merged**